src/gpt2/model.jl

```diff
@@ -0,0 +1,194 @@
+import Flux
+import ..Transformers
+
+export GPT2Model, GPT2Predictor
+
+
+struct GPT2Block{A<:Transformers.Basic.MultiheadAttention, L<:Flux.LayerNorm,
+                 F<:Transformers.Basic.PwFFN,
+                 D<:Flux.Dropout} <: Transformers.Basic.AbstractTransformer
+    pre_norm::L
+    attention::A
+    attn_norm::L
+    piecewise::F
+    dropout::D
+end
+
+Flux.@functor GPT2Block
+
+function GPT2Block(inputsize::Integer, num_heads::Integer, ffhiddensize::Integer;
+                   look_ahead::Bool=false, activation=Flux.gelu,
+                   p_dropout_attn=0.1f0)
+    if inputsize % num_heads != 0
+        throw(ArgumentError("`inputsize` not divisible by `num_heads`"))
+    end
+    GPT2Block(inputsize, num_heads, div(inputsize, num_heads), ffhiddensize;
+              look_ahead=look_ahead, activation=activation, p_dropout_attn=p_dropout_attn)
+end
+
+function GPT2Block(inputsize::Integer, heads::Integer, attnhiddensize::Integer,
+                   ffhiddensize::Integer;
+                   look_ahead::Bool=false, activation=Flux.gelu, p_dropout_attn=0.1f0)
+    GPT2Block(
+        Flux.LayerNorm(inputsize),
+        Transformers.Basic.MultiheadAttention(heads, inputsize, attnhiddensize, inputsize;
+                                              future=look_ahead,
+                                              pdrop=p_dropout_attn),
+        Flux.LayerNorm(inputsize),
+        Transformers.Basic.PwFFN(inputsize, ffhiddensize, activation),
+        Flux.Dropout(p_dropout_attn),
+    )
+end
+
+function (b::GPT2Block)(input::AbstractArray{T, N}, mask=nothing) where {T, N}
+    input = b.pre_norm(input)::typeof(input)
+    attn = b.attention(input, input, input; mask=mask)
+    attn = b.dropout(attn)
+    res_attn = input .+ attn
+    if N == 3
+        insize = size(res_attn)
+        res_attn = reshape(res_attn, insize[1], :)
+    end
+    res_attn = b.attn_norm(res_attn)::typeof(res_attn)
+    pwff = b.piecewise(res_attn)
+    pwff = b.dropout(pwff)
+    res_pwff = res_attn .+ pwff
+    if N == 3
+        res_pwff = reshape(res_pwff, :, Base.tail(insize)...)
+    end
+    res_pwff
```

Comment:
Could you check this implementation with the origin(openai) again? I only scan through, but it doesn't look correct.

Reply:
see here
https://github.com/openai/gpt-2/blob/0574c5708b094bfa0b0f6dfe3fd284d9a045acd9/src/model.py#L123-L130